sycl/plugins/unified_runtime/ur/adapters/native_cpu/context.cpp

```diff
@@ -0,0 +1,107 @@
+//===--------- context.cpp - Native CPU Adapter ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <memory>
+#include <tuple>
+
+#include "ur/ur.hpp"
+#include "ur_api.h"
+
+#include "common.hpp"
+#include "context.hpp"
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urContextCreate(uint32_t DeviceCount, const ur_device_handle_t *phDevices,
+                const ur_context_properties_t *pProperties,
+                ur_context_handle_t *phContext) {
+  std::ignore = pProperties;
+  UR_ASSERT(phDevices, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(phContext, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+
+  assert(DeviceCount == 1);
+
+  // TODO: Proper error checking.
+  auto ctx = new ur_context_handle_t_(*phDevices);
+  *phContext = ctx;
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urContextRetain(ur_context_handle_t hContext) {
+  std::ignore = hContext;
+  DIE_NO_IMPLEMENTATION
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urContextRelease(ur_context_handle_t hContext) {
+  delete hContext;
```

Comment:
The UR API objects really should implement reference counting, is this planned as future work?

This is one area where I think UR can provide some command shared code to assist with the implementation of object reference counting.

Reply:
Thanks for the comment.
Yes, this PR is mainly to get the main file structure and UR interfaces in place and "rebase" our initial WIP implementation on them. We'll implement these interfaces and integrate them closer with UR (to reuse as much as possible from UR) in subsequent PRs .